### ISSUE TITLE
[llvm_nvmf_fuzz] llvm_nvme_fuzz app hang up during the test

### ISSUE THREAD

**Issue by junwenx** [spdk/spdk#2481]
# Bug report
Hang up and raise an error when run below command

spdk/test/app/fuzz/llvm_nvme_fuzz/llvm_nvme_fuzz -m 0x1 -i 0 -F 'trtype:tcp adrfam:IPv4 subnqn:nqn.2016-06.io.spdk:cnode1 traddr:127.0.0.1 trsvcid:4420' -c /home/wenjun/spdk/test/nvmf/target/fuzz_json.conf -t 20 -D /tmp/llvm_fuzz0 -Z 0

## Expected Behavior

Fuzz normally

## Current Behavior

#22     NEW    cov: 1466 ft: 1991 corp: 20/629b lim: 320 exec/s: 1 rss: 168Mb L: 1/127 MS: 1 ChangeByte-
#24     NEW    cov: 1466 ft: 2049 corp: 21/656b lim: 320 exec/s: 1 rss: 168Mb L: 27/127 MS: 2 InsertRepeatedBytes-ChangeBit-
#25     NEW    cov: 1466 ft: 2072 corp: 22/734b lim: 320 exec/s: 1 rss: 169Mb L: 78/127 MS: 1 ShuffleBytes-
[2022-04-25 10:08:52.065820] nvme_ctrlr.c:3132:nvme_ctrlr_async_event_cb: *ERROR*: [nqn.2016-06.io.spdk:cnode1] Controller appears out-of-spec for asynchronous event request
handling.  Do not repost this AER.

## Steps to Reproduce

1. `rm -rf /tmp/llvm_fuzz*`
2. `cd spdk path` `CC=clang-12 CXX=clang++-12 ./configure --without-isal --with-fuzzer=/usr/lib64/clang/12/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a`

3. `make -j128`
4. `./test/nvmf/target/llvm_nvme_fuzz.sh 0 --time=20`
Error raise and system hang up.

 10:08:37       -- target/llvm_nvme_fuzz.sh@21 -- # /home/wenjun/spdk/test/app/fuzz/llvm_nvme_fuzz/llvm_nvme_fuzz -m 0x1 -i 0 -F 'trtype:tcp adrfam:IPv4 subnqn:nqn.2016-06.io.spdk:cnode1 traddr:127.0.0.1 trsvcid:4420' -c /home/wenjun/spdk/test/nvmf/target/fuzz_json.conf -t 20 -D /tmp/llvm_fuzz0 -Z 0
[2022-04-25 10:08:37.268500] Starting SPDK v22.05-pre git sha1 494eb6e58 / DPDK 21.11.0 initialization...
[2022-04-25 10:08:37.269639] [ DPDK EAL parameters: [2022-04-25 10:08:37.269670] nvme_fuzz [2022-04-25 10:08:37.269686] -c 0x1 [2022-04-25 10:08:37.269698] --log-level=lib.eal:6 [2022-04-25 10:08:37.269711] --log-level=lib.cryptodev:5 [2022-04-25 10:08:37.269723] --log-level=user1:6 [2022-04-25 10:08:37.269736] --base-virtaddr=0x200000000000 [2022-04-25 10:08:37.269749] --match-allocations [2022-04-25 10:08:37.269767] --file-prefix=spdk0 [2022-04-25 10:08:37.269783] --proc-type=auto [2022-04-25 10:08:37.269799] ]
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
[2022-04-25 10:08:37.358288] app.c: 603:spdk_app_start: *NOTICE*: Total cores available: 1
[2022-04-25 10:08:37.458119] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 0
[2022-04-25 10:08:37.482950] accel_engine.c: 548:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
[2022-04-25 10:08:37.577466] tcp.c: 619:nvmf_tcp_create: *NOTICE*: *** TCP Transport Init ***
[2022-04-25 10:08:37.594132] tcp.c: 822:nvmf_tcp_listen: *NOTICE*: *** NVMe/TCP Target Listening on 127.0.0.1 port 4420 ***
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1094662269
INFO: Loaded 1 modules   (30178 inline 8-bit counters): 30178 [0x820183, 0x827765),
INFO: Loaded 1 PC tables (30178 PCs): 30178 [0x827768,0x89d588),
INFO:        0 files found in /tmp/llvm_fuzz0
INFO: A corpus is not provided, starting from an empty corpus
#2      INITED cov: 1417 ft: 1418 corp: 1/1b exec/s: 2 rss: 166Mb
#3      NEW    cov: 1422 ft: 1476 corp: 2/2b lim: 320 exec/s: 3 rss: 167Mb L: 1/1 MS: 1 CopyPart-
#4      pulse  cov: 1422 ft: 1538 corp: 2/2b lim: 320 exec/s: 2 rss: 167Mb
#4      NEW    cov: 1428 ft: 1538 corp: 3/3b lim: 320 exec/s: 2 rss: 167Mb L: 1/1 MS: 1 ChangeByte-
#5      NEW    cov: 1428 ft: 1615 corp: 4/4b lim: 320 exec/s: 2 rss: 167Mb L: 1/1 MS: 1 ChangeBit-
#6      NEW    cov: 1428 ft: 1631 corp: 5/5b lim: 320 exec/s: 2 rss: 167Mb L: 1/1 MS: 1 ShuffleBytes-
#7      NEW    cov: 1428 ft: 1657 corp: 6/6b lim: 320 exec/s: 2 rss: 167Mb L: 1/1 MS: 1 CopyPart-
#8      pulse  cov: 1428 ft: 1673 corp: 6/6b lim: 320 exec/s: 2 rss: 167Mb
#8      NEW    cov: 1428 ft: 1673 corp: 7/8b lim: 320 exec/s: 2 rss: 167Mb L: 2/2 MS: 1 CopyPart-
#9      NEW    cov: 1428 ft: 1687 corp: 8/9b lim: 320 exec/s: 2 rss: 167Mb L: 1/2 MS: 1 ChangeByte-
#10     NEW    cov: 1428 ft: 1730 corp: 9/10b lim: 320 exec/s: 2 rss: 167Mb L: 1/2 MS: 1 ChangeByte-
#11     NEW    cov: 1428 ft: 1748 corp: 10/11b lim: 320 exec/s: 2 rss: 167Mb L: 1/2 MS: 1 ChangeBit-
#12     NEW    cov: 1428 ft: 1774 corp: 11/12b lim: 320 exec/s: 2 rss: 167Mb L: 1/2 MS: 1 ShuffleBytes-
        NEW_FUNC[1/5]: 0x461410 in fuzz_admin_command /home/wenjun/spdk/test/app/fuzz/llvm_nvme_fuzz/llvm_nvme_fuzz.c:70
        NEW_FUNC[2/5]: 0x462350 in nvme_fuzz_cpl_cb /home/wenjun/spdk/test/app/fuzz/llvm_nvme_fuzz/llvm_nvme_fuzz.c:490
#13     NEW    cov: 1452 ft: 1813 corp: 12/138b lim: 320 exec/s: 1 rss: 168Mb L: 126/126 MS: 1 InsertRepeatedBytes-
#14     NEW    cov: 1455 ft: 1875 corp: 13/265b lim: 320 exec/s: 2 rss: 168Mb L: 127/127 MS: 1 CrossOver-
#15     NEW    cov: 1455 ft: 1897 corp: 14/391b lim: 320 exec/s: 1 rss: 168Mb L: 126/127 MS: 1 ShuffleBytes-
#16     pulse  cov: 1455 ft: 1907 corp: 14/391b lim: 320 exec/s: 2 rss: 168Mb
#16     NEW    cov: 1455 ft: 1907 corp: 15/392b lim: 320 exec/s: 2 rss: 168Mb L: 1/127 MS: 1 ChangeBit-
#18     NEW    cov: 1455 ft: 1945 corp: 16/393b lim: 320 exec/s: 2 rss: 168Mb L: 1/127 MS: 2 InsertByte-EraseBytes-
#19     NEW    cov: 1460 ft: 1955 corp: 17/471b lim: 320 exec/s: 1 rss: 168Mb L: 78/127 MS: 1 InsertRepeatedBytes-
#20     NEW    cov: 1462 ft: 1970 corp: 18/550b lim: 320 exec/s: 1 rss: 168Mb L: 79/127 MS: 1 InsertByte-
#21     NEW    cov: 1465 ft: 1981 corp: 19/628b lim: 320 exec/s: 1 rss: 168Mb L: 78/127 MS: 1 CrossOver-
#22     NEW    cov: 1466 ft: 1991 corp: 20/629b lim: 320 exec/s: 1 rss: 168Mb L: 1/127 MS: 1 ChangeByte-
#24     NEW    cov: 1466 ft: 2049 corp: 21/656b lim: 320 exec/s: 1 rss: 168Mb L: 27/127 MS: 2 InsertRepeatedBytes-ChangeBit-
#25     NEW    cov: 1466 ft: 2072 corp: 22/734b lim: 320 exec/s: 1 rss: 169Mb L: 78/127 MS: 1 ShuffleBytes-
[2022-04-25 10:08:52.065820] nvme_ctrlr.c:3132:nvme_ctrlr_async_event_cb: *ERROR*: [nqn.2016-06.io.spdk:cnode1] `Controller appears out-of-spec for asynchronous event request
handling.  Do not repost this AER.`

## Context (Environment including OS version, SPDK version, etc.)

SPDK version: 494eb6e58b8e72762d0a2127f2657926298e2523
OS: `fedora34`
kernel: `5.16.18-100.fc34.x86_64`
Clang: clang-12

**Comment by Comphix** [comment]
Looks like target side caught some error from the nvme_fuzz app. Is there any log from nvme_fuzz side?

https://github.com/spdk/spdk/pull/25 NEW cov: 1466 ft: 2072 corp: 22/734b lim: 320 exec/s: 1 rss: 169Mb L: 78/127 MS: 1 ShuffleBytes-
[2022-04-25 10:08:52.065820] nvme_ctrlr.c:3132:nvme_ctrlr_async_event_cb: ERROR: [nqn.2016-06.io.spdk:cnode1] Controller appears out-of-spec for asynchronous event request
handling. Do not repost this AER.

**Comment by wanqunintel** [comment]
@junwenx , thanks to reporting this out, i also tried the step on my environment and it passed the test. my kernel is: 5.11.12-300.fc34.x86_64. same spdk commit as you.  Clang12. Would you share me the seed file so that i can reproduce this issue the same as your input?

**Comment by junwenx** [comment]
@wanqunintel , you may need to rebase the newest code, and the seed file will mail to you. You may need to execute make clean, re-config and make -j128 to compile code, and then do the test.

**Comment by Bianyfan** [comment]
It seems to be related with recently code change, I rebase to the newest code and reproduce this error. I may make a double check.

**Comment by Bianyfan** [comment]
Just update the status, I tested the early version like 6301f89 that was on April 1st. Actually, It occurred the same issue with lower frequency, which has been confirmed with wenjun.  In addition, I checked the fuzz app, there was an infinite loop during the run_cmds() function.

**Comment by changpe1** [comment]
There is no way that the llvm_nvme_fuzz application test cases to send the AER command, but when doing `spdk_nvme_connect`, the NVMe driver will send AER command during initialization, this issue may not related with llvm_nvme_fuzz test, we can test the case that  attach and detach a controller many times to see this issue happen or not.

**Comment by Bianyfan** [comment]
Hi, @changpe1 thanks for your comments. Actually, I have a regression and find a related commit: https://review.spdk.io/gerrit/c/spdk/spdk/+/12017. In short, during fuzz side, we submit a command and set a cb to update status and break the loop when it complete. But in this procedure, we just mark it and fail this request without sending to the target layer. That means in fuzz side, it's no way to update cb status and it will hang up in the loop. I raise a patch(https://review.spdk.io/gerrit/c/spdk/spdk/+/12402) to test and it works under wenjun's test. In addition, if we drop command in other scenarios and fuzz side may hang up again.

**Comment by jimharris** [comment]
> Hi, @changpe1 thanks for your comments. Actually, I have a regression and find a related commit: https://review.spdk.io/gerrit/c/spdk/spdk/+/12017. In short, during fuzz side, we submit a command and set a cb to update status and break the loop when it complete. But in this procedure, we just mark it and fail this request without sending to the target layer. That means in fuzz side, it's no way to update cb status and it will hang up in the loop. I raise a patch(https://review.spdk.io/gerrit/c/spdk/spdk/+/12402) to test and it works under wenjun's test. In addition, if we drop command in other scenarios and fuzz side may hang up again.

Hi Yifan,

I posted a comment on the patch.  This patch may get this fuzz case to stop hanging, but am not sure it is valid.  We need to understand more about the exact sequence of commands that causes this hang.

-Jim

**Comment by Bianyfan** [comment]
Hi @jimharris , I reproduced this issue based on the crash with your patch. Exactly, it's related with the fused operation. I printed the fused field number was 2 and opc ==SPDK_NVME_OPC_CREATE_IO_SQ. In addition, I used number 1 for fused field with the same opc or different opc like SPDK_NVME_OPC_DELETE_IO_SQ to test and it hung up again. So I think we may not deal correctly with fused operation. BTW, I l think Fuzzer 0 is for random testing of admin command, we needn't to set these commands together and just raise this patch to remove it. https://review.spdk.io/gerrit/c/spdk/spdk/+/12421

**Comment by changpe1** [comment]
> Hi @jimharris , I reproduced this issue based on the crash with your patch. Exactly, it's related with the fused operation. I printed the fused field number was 2 and opc ==SPDK_NVME_OPC_CREATE_IO_SQ. In addition, I used number 1 for fused field with the same opc or different opc like SPDK_NVME_OPC_DELETE_IO_SQ to test and it hung up again. So I think we may not deal correctly with fused operation. BTW, I l think Fuzzer 0 is for random testing of admin command, we needn't to set these commands together and just raise this patch to remove it. https://review.spdk.io/gerrit/c/spdk/spdk/+/12421

@Bianyfan 
Fuzz test to skip the FUSED bits is OK, but we should get it merged after the root cause of this issue.

Are you saying that with `OPC=SPDK_NVME_OPC_CREATE_IO_SQ and FUSED=2` can reproduce this issue?

**Comment by Bianyfan** [comment]
> hi @changpe1 , yes a simple way to reproduce is just to set cmd->cmd.fuse = 2 and cmd->cmd.opc = SPDK_NVME_OPC_CREATE_IO_SQ at the last of fuzz_admin_command. Then start a fuzz with this command "./test/nvmf/target/llvm_nvme_fuzz.sh 0 --time=60"

**Comment by jimharris** [comment]
I root caused this test failure today.  It found a real bug in the misordered fused command handling on the target.

https://review.spdk.io/gerrit/c/spdk/spdk/+/12452 is the fix.  The commit message describes the problem:

```
nvmf: set cid when failing a misordered fused command

If we don't set the cid before failing the misordered
command, we use some other random cid, causing the
initiator to think the wrong command was completed.

Fixes #2481.

For this issue, the target was completing a
previously submitted AER, not the fuzzed fused
command.  The initiator would then submit another
AER to replace the completed one, but the target
complained that the initiator sent too many AERs
since the target didn't really know it had completed
an AER so hadn't adjusted its num_aer count.
```

This patch set also includes a fix to force FUSE bits on fuzzer 0 (admin cmd fuzzing) to zero always.  If the last fuzzed command has FUSE=1, the keep alive will eventually kick that command to (failed) completion - but that won't happen on other transports.

This patch set also includes a new feature for the fuzzer, where the data file generated by LLVM for a previous crash/hang can be used as a single-shot test run for purposes of reproduction.  I used this feature as part of the debug on this issue.  User can specify -N <data file> and then llvm_nvme_fuzz won't do any fuzzing, it will just run a single iteration using input data from the specified file.

**Comment by junwenx** [comment]
@jimharris @changpe1 @Bianyfan @Comphix Thank you for the help, I tried to pull Jim patch on local, and run pass, it looks like fix the issue. Thanks!
